Ullaakut/gonvey
# Add prometheus & grafana to docker-compose.yml + Preset drashboard (#5)

## Goal of this PR

This PR adds:
* A Prometheus instance running alongside Gonvey in the `docker-compose.yml` file
* A Grafana instance with a preconfigured source and dashboard to integrate with Prometheus
* Changes in Gonvey to push metrics to Prometheus in real time

Prometheus and Grafana both have persistency through the use of docker volumes.

The tracked metrics are:

* The basic `promhttp` metrics
* `http_response_time`
* `http_requests_count` (with labels:)
  * `http_method`
  * `http_request_uri`
  * `endpoint`
* `http_remote_addr`
* `http_response_code`

Fixes #5

## How to test it

* `docker-compose up`
* Visit `localhost:3000` in your favorite browser and login with `admin` / `admin`. You can skip the password change.
* Go on the `Gonvey` dashboard (it might take a few seconds before it is plugged with the Prometheus source
* Make some requests on the proxy
  * `curl -v "http://0.0.0.0:8888/test/posts"`
  * `curl -v "http://0.0.0.0:8888/bloggo/posts"`
  * `curl -v "http://0.0.0.0:8888/test/invalid/path"`
  * `curl -v "http://0.0.0.0:8888/bloggo/invalid/path"`
  * Ideally do a while/for loop to simulate multiple user requests
* Verify that the dashboard displays correct information like such

<img width="1700" alt="screenshot 2018-09-23 at 13 02 37" src="https://user-images.githubusercontent.com/6976628/45927349-1ff1a400-bf32-11e8-8f6f-3138bc7b6f4d.png">

* `docker-compose restart`
* Access Grafana on `localhost:3000` again, and verify that the data from before restarting the containers is still there.
* Verify that the information displayed in the graphs makes sense.
